### PR TITLE
[feat] 2/2 Add auto-id for trainer.predict output

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds `auto-id` on top of Trainer.predict

Fixes # (issue) <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/supporters.py [hunk 1]
```diff
@@ -118,27 +122,278 @@ def mean(self):
...
+        })
+
+        missing_kwargs = valid_kwargs.difference(LightningBatchSamplerWrapper.SKIP_VALID_KEYS).difference(dl_args)
+        if missing_kwargs:
+            dataloader_cls_name = dataloader.__class__.__name__
+            rank_zero_warn(
+                f"Trying to replace to wrap your BatchSampler in your {dataloader_cls_name} dataloader."
+                "This would fail as your DataLoader doesn't expose as attributes all its __init__ parameters. "
+                f"Missing attributes are {missing_kwargs} \n"
+                "HINT: use Trainer(enable_predict_auto_id=False) and provide your own id."
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/supporters.py:187]: i don't know what this does....
can you rename the flag to make it clear? or add docstrings or a much clearer error message?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -291,11 +292,18 @@ def __init__(
                 In 'max_size_cycle' mode, the trainer ends one epoch when the largest dataset is traversed,
                 and smaller datasets reload when running out of their data. In 'min_size' mode, all the datasets
                 reload when reaching the minimum length of datasets.
+
+            enable_predict_auto_id: Explicitly enables or disables sampler replacement.
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/trainer.py:296]: what does a sampler have to do with predict?

**Review by williamFalcon — CHANGES_REQUESTED** [review]

**Comment by Borda** [comment]
@tchaton how is it going here? probably need some large rebase :]
